Surrogate ML model update

lpottier opened the pull request:
The surrogate model can be now updated at runtime when a RabbitMQ message following this template is received by AMSlib:
```
UPDATE:/surrogate/ml/path
``` 

The message has to be sent on the `exchange` and `routing key` specified in the JSON config.
```json
{
  "db" : {
    "dbType" : "rmq",
      "rmq_config" : {
        "rabbitmq-name": "",
        "rabbitmq-password": "",
        "rabbitmq-user": "",
        "rabbitmq-vhost": "",
        "service-port": 0,
        "service-host": "",
        "rabbitmq-cert": "",
        "rabbitmq-outbound-queue": "",
        "rabbitmq-exchange": "",
        "rabbitmq-routing-key": ""
    },
    "update_surrogate": true
  },
  "ml_models" : {
  }
}
```

Note that `rabbitmq-inbound-queue` has been deprecated and is now replaced by `rabbitmq-exchange` and `rabbitmq-routing-key`. A new field called `update_surrogate` has been added which controls whether AMS will update its surrogate.

lpottier committed:
Moved most implementations of RMQ to its own .cpp file

Signed-off-by: Loic Pottier <pottier1@llnl.gov>

lpottier committed:
Added ML model update.
The surrogate model will be updated when a message following the template:
    UPDATE:/surrogate/ml/path will be received from RMQ.
The message has to be sent on the exchange and routing key specified in the JSON config.
{
    rabbitmq-exchange: "exchange",
    rabbitmq-routing-key: "key"
}

Signed-off-by: Loic Pottier <pottier1@llnl.gov>

koparasy commented on src/AMSlib/AMS.cpp:
```diff
@@ -485,6 +491,8 @@ class AMSWrap
 public:
   AMSWrap() : memManager(ams::ResourceManager::getInstance())
   {
+    rId = get_rank_id();
```
I am a little worried about this here. The `env` variables I am using are not always supported and depend on the job scheduler (on a non-workflow execution we don't have any control over). I skimmed over the code and this seems to be mostly used as an identifier in Logging messages right?

Which part of rId is necessary for instantiating the db before application execution?

lpottier replied:
I agree this solution is not the best. The problem here is that I cannot use `MPI_Comm_rank` in RMQ part because AMSWrap is static and will get instantiated before `MPI_Init` is being called resulting an an MPI error.

I use the `rId` for three things
1. For logging
2. To embed the MPI rank in AMSMessage
3. To name uniquely inbound queues (queue that receive messages from the outside). Each of these should be uniquely named otherwise some ranks will not receive messages with the new ML model (for example).

I am working on a solution where I use MPI to get the rank before sending the `AMSMessage` (at that point MPI is initialized) and I use a random generator to uniquely named RMQ queue in the `RMQConsumerHandler`. For the logging, the `rId` will be the random number, so it's not perfect but it's okay I think.

koparasy replied:
Ok. I will go over the implementation and see if I get any idea. (1) is not that important and for 1 you can use the current id. 
2. Should be feasible. Since the DBManager creates files in the form `data_<rid>.<suffix>`. I don't recall the details of AMSMessage. So maybe that is too late in the pipeline.
3. Can we build a lazy queue creation mechanism? That takes the first time we call the `getLatestModel` ?

lpottier replied:
I pushed a new version without `rId` by implementing 1. and 2.

It would be complicated (but not impossible) to build a lazy queue creation mechanism. The problem is that we have to create the queue before any messages is being sent to AMSlib, otherwise we will lose these messages (if you send a message to a queue that does not exist then this message is lost).
If we know the names of the queues beforehand, we could create them on the Python side for example. If these queues exist then messages sent to them will be kept.

koparasy commented on src/AMSlib/CMakeLists.txt:
```diff
@@ -17,6 +17,9 @@ if (WITH_HDF5)
 
+if (WITH_RMQ)
+  list(APPEND AMS_LIB_SRC wf/rmqdb.cpp)
+endif()
```
Nice!

koparasy commented on src/AMSlib/include/AMS.h:
```diff
@@ -111,7 +111,7 @@ void AMSDestroyExecutor(AMSExecutor executor);
 void AMSSetAllocator(AMSResourceType resource, const char *alloc_name);
 const char *AMSGetAllocatorName(AMSResourceType device);
-void AMSConfigureFSDatabase(AMSDBType db_type, const char *db_path);
+void AMSConfigureFSDatabase(uint64_t rId, AMSDBType db_type, const char *db_path);
```
This goes back to the rId comment. I am not sure this should be necessary.

lpottier replied:
See reply above

koparasy commented on src/AMSlib/wf/debug.h:
```diff
@@ -25,8 +25,11 @@ void memUsage(double& vm_usage, double& resident_set);
               "[AMS:%s:%s] ",                      \
               ams::util::getVerbosityKey(vl),      \
               GETNAME(id));                        \
+      fflush(ams::util::out(vl));                  \
```
is `fflush` necessary? I am guessing this is because multi-thread execution. But it may be too high of a penalty. Let me know which use case you had in mind when you started flushing?

lpottier replied:
Yes I wanted to fix the printing which can be messed up and I tried a few things but it's more complicated than just flush so I gave up. But I forgot to remove the fflush. Good catch.

koparasy commented on src/AMSlib/wf/workflow.hpp:
```diff
@@ -165,6 +165,18 @@ class AMSWorkflow
     Store(num_elements, mInputs, outputs, predicate);
   }
 
+  bool updateModel()
+  {
+    if (!DB) return false;
+    bool local = DB->updateModel();
+#ifdef __ENABLE_MPI__
+    bool global = false;
+    MPI_Allreduce(&local, &global, 1, MPI_CXX_BOOL, MPI_LAND, comm);
+    return global;
+#else
+    return local;
+#endif
+  }
```
This is a global sync right? Happening in every timestep?

lpottier replied:
Yes. We could implement some relaxations:

1. We could check globally **only** if rank 0 has received something (here if `local` is true)
2. We could check every X timesteps

koparasy commented on src/AMSlib/wf/workflow.hpp:
```diff
@@ -318,9 +330,16 @@ class AMSWorkflow
 
-    if (DB && DB->updateModel()) {
-      UQModel->updateModel("");
+    CALIPER(CALI_MARK_BEGIN("UPDATEMODEL");)
+    if (updateModel()) {
+      auto model = DB->getLatestModel();
+      CINFO(Workflow,
+            rId == 0,
+            "Updating surrogate model with %s",
+            model.c_str())
+      UQModel->updateModel(model);
```
In a case where multiple models exist in the simulation how would we handle this? Food for thought ...

lpottier replied:
We will have to discuss that use case indeed, and how we select the correct model to use. We could had a model tag in the message that AMSlib receive with the model name and update the right model accordingly.

koparasy reviewed (commented):
Hey. Thank you for the PR. This has been a lot of work. Thank you also for splitting rmq into a separate cpp file. I have some high level comments. I still haven't gone through the implementation of "src/AMSlib/wf/rmqdb.cpp" will try to do it tomorrow morning. Can you please look at this general feedback?

lpottier committed:
Based on PR review:
- removed rId requirement by moving to random generator queue ID
- When publishing messages, if MPI is initialized, we will send AMSMessage with correct MPI rank

Signed-off-by: Loic Pottier <pottier1@llnl.gov>

lpottier committed:
Added scripts to debug/test RabbitMQ in tools/

Signed-off-by: Loic Pottier <pottier1@llnl.gov>

lpottier committed:
Updated RMQ script tools with a script to send AMS-compatible binary messages.

Signed-off-by: Loic Pottier <pottier1@llnl.gov>

koparasy commented on src/AMSWorkflow/ams/rmq.py:
```diff
@@ -55,20 +56,23 @@ def endianness(self) -> str:
         # float or double
-        msg_format = f"{header_format}{data.size}{dt}"
-        return struct.pack(msg_format, *header_content, *data)
+        msg_format = f"{header_format}{domain_name_size}s{data.size}{dt}"
```
What is the s between the domain_name_size and data.size?

lpottier replied:
It indicates a string here (cf [Python struct format](https://docs.python.org/3/library/struct.html#format-characters))

koparasy replied:
Thnx

koparasy commented on src/AMSlib/wf/basedb.hpp:
The definition of the AMSHeader should be moved into the cpp file.

lpottier replied:
Done

koparasy commented on src/AMSlib/wf/basedb.hpp:
```diff
@@ -1065,61 +973,135 @@ class AMSMessage
+  AMSMessageInbound(AMSMessageInbound&&) = default;
+  AMSMessageInbound& operator=(AMSMessageInbound&&) = default;
+
+  AMSMessageInbound(uint64_t id,
+                    uint64_t rId,
+                    std::string body,
+                    std::string exchange,
+                    std::string routing_key,
+                    bool redelivered)
+      : id(id),
+        rId(rId),
+        body(std::move(body)),
+        exchange(std::move(exchange)),
+        routing_key(std::move(routing_key)),
+        redelivered(redelivered){};
```
Move definition to cpp file

lpottier replied:
Done

koparasy commented on src/AMSlib/wf/basedb.hpp:
```diff
@@ -1065,61 +973,135 @@ class AMSMessage
-  std::string _queue;
-  /** @brief Queue that contains all the messages received on receiver queue */
-  std::shared_ptr<std::vector<inbound_msg>> _messages;
+  // /** @brief main channel used to send data to the broker */
+  // std::shared_ptr<AMQP::TcpChannel> _channel;
+  // /** @brief AMQP reliable channel (wrapper of classic channel with added functionalities) */
+  // std::shared_ptr<AMQP::Reliable<AMQP::Tagger>> _rchannel;
```
Do we need these?

lpottier replied:
No, good catch

koparasy commented on src/AMSlib/wf/basedb.hpp:
```diff
@@ -1175,111 +1130,10 @@ class RMQConsumerHandler final : public AMQP::LibEventHandler
-    DBG(RMQConsumerHandler,
-        "[rank=%d] Secured TLS connection has been established.",
-        _rank)
+    DBG(RMQHandler, "[r%d] Secured TLS connection has been established.", _rId)
```
to cpp file?

lpottier replied:
Done

koparasy commented on src/AMSlib/wf/basedb.hpp:
```diff
@@ -1289,7 +1143,7 @@ class RMQConsumerHandler final : public AMQP::LibEventHandler
   virtual void onClosed(AMQP::TcpConnection* connection) override
   {
-    DBG(RMQConsumerHandler, "[rank=%d] Connection is closed.\n", _rank)
+    DBG(RMQHandler, "[r%d] Connection is closed.", _rId)
```
to cpp file

lpottier replied:
Done

koparasy commented on src/AMSlib/wf/basedb.hpp:
```diff
@@ -1302,26 +1156,120 @@ class RMQConsumerHandler final : public AMQP::LibEventHandler
   {
-    //  add your own implementation, like cleanup resources or exit the application
-    DBG(RMQConsumerHandler, "[rank=%d] Connection is detached.\n", _rank)
+    // Different queue name for each consumer (must be unique)
+    // In RMQ queues must be named excplicitly if we want to keep messages
+    // if unamed queue used, then messages sent before that queue is set up will be lost forever
+    // Important: queue names cannot be > 255 bytes
+    _queue = "amslib-" + std::to_string(rId);
+    CWARNING(RMQConsumerHandler,
+             _queue.length() > 255,
+             "Queue name \"%s\" is too long (%d)",
+             _queue.c_str(),
+             _queue.length())
```
Can you add a NOTE/TODO comment. The fact that we expect to have a queue per rank may have performance implications.

lpottier replied:
Done.

koparasy commented on src/AMSlib/wf/basedb.hpp:
```diff
@@ -1331,106 +1279,106 @@ class RMQConsumer
 
   /**
-   * @brief Start the underlying I/O loop (blocking call)
+   *  @brief Start the underlying I/O loop (blocking call)
```
Move function definitions to cpp file

lpottier replied:
Done

koparasy commented on src/AMSlib/wf/tmp-publish_state.cpp:
The file name is weird. DId you mean to push this?

lpottier replied:
No, good catch, This was an old implementation. Removed.

koparasy commented on src/AMSlib/wf/workflow.hpp:
```diff
@@ -121,7 +121,11 @@ class AMSWorkflow
 
     if (appDataLoc == AMSResourceType::AMS_HOST) {
+#ifdef __ENABLE_MPI__
+      return DB->store(num_elements, inputs, outputs, comm, predicate);
```
It is unclear to me why comm is required when storing data. I will look closer in the code. It feels though that we start adding a secondary "meaning" to the store command.

lpottier replied:
Yes it is not the cleaner implementation. It leads to a lot of conditional code which I don't like. The only reason why I pass down the MPI comm to encode the actual MPI rank in AMSMessage. I considered an alternative implementation where I would use the random ID used for logging in AMSMessage, the advantage is a cleaner implementation because `store` would not need any MPI_Comm but the AMSMessage will not have the actual MPI rank encoded (if that matters). We could also choose to not encode MPI rank in AMSMessage.

Let me know your thoughts about it.

koparasy replied:
Wouldn't it be enough to pass the rank-id when constructing the DB object itself?

koparasy replied:
[Here](https://github.com/LLNL/AMS/blob/features/mlmodel-update/src/AMSlib/wf/workflow.hpp#L210) you have the rId. Over [here](https://github.com/LLNL/AMS/blob/features/mlmodel-update/src/AMSlib/wf/workflow.hpp#L227)  we pass it to the DB manager. I can see also [here](https://github.com/LLNL/AMS/blob/features/mlmodel-update/src/AMSlib/wf/basedb.hpp#L1996) that we forward the rank id to RMQ.

lpottier replied:
You are right I missed the obvious solution... I refactored everything and removed `comm` from `store` methods!

koparasy reviewed (changes requested):
I added some minor comments. Mainly to move definitions from the header to the cpp file. It is still unclear to me some of the concepts of the update model and the model receiver. We should discuss.

lpottier committed:
Moved RMQ implementation to corresponding .cpp file
Deleted useless code
We use random queues generated by RMQ instead of doing it manually in AMSlib

Signed-off-by: Loic Pottier <pottier1@llnl.gov>

lpottier commented:
@koparasy I have addressed your comments, let me know when you want to discuss.

koparasy commented on src/AMSlib/wf/workflow.hpp:
```diff
@@ -162,9 +170,21 @@ class AMSWorkflow
+    store(num_elements, mInputs, outputs, predicate);
   }
 
+  bool updateModel()
```
One more comment :D . Can we make updateModel an optional dynamic capability? Se by default to false?

As an initial step, I would add a field on DBManager set it false and optionally toggle it to true when we set some value on the json file.

lpottier replied:
Good idea, I added a flag (by default no model update will be performed). I am working on the JSON flag

koparasy replied:
Thnx!

lpottier replied:
Done. I've added a new field that controls the update:
```json
  "db" : {
    "dbType" : "rmq",
    "rmq_config" : {
    },
    "update_surrogate": true
  },
```

lpottier committed:
Removed MPI communicator from DB store() based on PR discussions
added flags to updateModel to make it conditional

Signed-off-by: Loic Pottier <pottier1@llnl.gov>

koparasy reviewed (approved):
Thank you!

lpottier committed:
Added JSON option db:update_surrogate (boolean) to control whether we update surrogate

Signed-off-by: Loic Pottier <pottier1@llnl.gov>
